### PR TITLE
First pass at supporting optimized pagination.

### PR DESCRIPTION
**Pull request by mattdaw**
Requires code not merged into Shotgun master just yet...

### CODE REVIEW

**Comment by mattdaw** [comment]
@pboucher The equivalent server-side code is here: https://github.com/shotgunsoftware/shotgun/pull/4424/files... Should be in master soon.

**Comment by mattdaw** [comment]
Pagination changes landed in master...

File: shotgun_api3/shotgun.py [hunk 1]
```diff
@@ -833,37 +841,57 @@ def find(self, entity_type, filters, fields=None, order=None,
                                                  include_archived_projects,
                                                  additional_filter_presets)
 
+        if self.server_caps.version and self.server_caps.version >= (3, 3, 0):
```
Comment (pboucher) [inline shotgun_api3/shotgun.py]: Can we use a `server_caps` `ensure_X` method here (without `raise_hell = True`) to standardize?

Comment (mattdaw) [reply shotgun_api3/shotgun.py]: Sure

**Pull request merged**